codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -243,66 +246,43 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                                 "Stream %d in unexpected state: %s", stream.id(), stream.state()));
                 }
             }
-
-            if (lastDataWrite != null && !endOfStream) {
-                throw new IllegalStateException(
-                        "Sending non-trailing headers after data has been sent for stream: "
-                                + streamId);
+            ChannelFuture future;
+            if (openStream) {
```

Comment:
Seems like there are a few things going on here which have inter-dependencies:
1. Managing stream state.
2. Writing (or queuing) headers.

The convention thus far has been to manage stream state has thus far been here in `Http2ConnectionEncoder`.  Are you suggesting that we move all (or some) of the stream state management to the flow controller?  It may be a bit awkward to check the stream state here, detect a transition needs to occur, and then rely on another class to make that transition.  Can you provide more detail with your suggestion?